File: workers/storage/local_adapter.py
"""Local filesystem storage adapter."""

from __future__ import annotations

import os
from pathlib import Path
from typing import List

from workers.core.errors import PermanentError, RetryableError
from workers.storage.adapter import StorageAdapter


class LocalStorageAdapter(StorageAdapter):
    """Storage adapter for local filesystem."""

    def __init__(self, base_path: str):
        """
        Initialize local storage adapter.
        
        Args:
            base_path: Base directory for storage (e.g., "/shared")
        """
        self.base_path = Path(base_path)
        if not self.base_path.exists():
            raise ValueError(f"Base path does not exist: {base_path}")

    def _resolve_path(self, object_key: str) -> Path:
        """Convert object key to absolute filesystem path."""
        # Remove leading slashes
        clean_key = object_key.lstrip("/\\")
        return self.base_path / clean_key

    def read_file(self, object_key: str) -> bytes:
        """Read file from local filesystem."""
        file_path = self._resolve_path(object_key)
        try:
            with open(file_path, "rb") as handle:
                return handle.read()
        except FileNotFoundError as exc:
            raise PermanentError(f"File not found: {object_key}") from exc
        except PermissionError as exc:
            raise PermanentError(f"Permission denied: {object_key}") from exc
        except Exception as exc:
            raise RetryableError(f"Failed to read file: {object_key}") from exc

    def write_file(self, object_key: str, content: bytes) -> None:
        """Write file to local filesystem."""
        file_path = self._resolve_path(object_key)
        try:
            # Create parent directories
            file_path.parent.mkdir(parents=True, exist_ok=True)
            with open(file_path, "wb") as handle:
                handle.write(content)
        except PermissionError as exc:
            raise PermanentError(f"Permission denied: {object_key}") from exc
        except Exception as exc:
            raise RetryableError(f"Failed to write file: {object_key}") from exc

    def file_exists(self, object_key: str) -> bool:
        """Check if file exists on local filesystem."""
        file_path = self._resolve_path(object_key)
        return file_path.exists() and file_path.is_file()

    def get_file_size(self, object_key: str) -> int:
        """Get file size from local filesystem."""
        file_path = self._resolve_path(object_key)
        try:
            return os.path.getsize(file_path)
        except FileNotFoundError as exc:
            raise PermanentError(f"File not found: {object_key}") from exc
        except Exception as exc:
            raise RetryableError(f"Failed to get file size: {object_key}") from exc

    def list_files(self, prefix: str) -> List[str]:
        """List files with given prefix on local filesystem."""
        prefix_path = self._resolve_path(prefix)
        if not prefix_path.exists():
            return []
        
        files = []
        if prefix_path.is_file():
            # Prefix is a file
            files.append(prefix)
        elif prefix_path.is_dir():
            # Prefix is a directory - list all files recursively
            for path in prefix_path.rglob("*"):
                if path.is_file():
                    # Convert back to relative key
                    relative = path.relative_to(self.base_path)
                    files.append(str(relative).replace("\\", "/"))
        
        return files

    def delete_file(self, object_key: str) -> None:
        """Delete file from local filesystem."""
        file_path = self._resolve_path(object_key)
        try:
            if file_path.exists():
                file_path.unlink()
        except Exception as exc:
            raise RetryableError(f"Failed to delete file: {object_key}") from exc

File: src/storage/localAdapter.js
/**
 * Local filesystem storage adapter for Node.js
 */

import fs from 'fs/promises';
import path from 'path';
import { StorageAdapter } from './adapter.js';

export class LocalStorageAdapter extends StorageAdapter {
    constructor(basePath) {
        super();
        this.basePath = basePath;
    }

    _resolvePath(objectKey) {
        // Remove leading slashes
        const cleanKey = objectKey.replace(/^[/\\]+/, '');
        return path.join(this.basePath, cleanKey);
    }

    async readFile(objectKey) {
        const filePath = this._resolvePath(objectKey);
        try {
            return await fs.readFile(filePath);
        } catch (error) {
            if (error.code === 'ENOENT') {
                throw new Error(`File not found: ${objectKey}`);
            }
            throw new Error(`Failed to read file: ${objectKey} - ${error.message}`);
        }
    }

    async writeFile(objectKey, content) {
        const filePath = this._resolvePath(objectKey);
        console.log(`[LocalAdapter] Writing file to local filesystem`);
        console.log(`[LocalAdapter]   Object key: ${objectKey}`);
        console.log(`[LocalAdapter]   Full path: ${filePath}`);
        console.log(`[LocalAdapter]   Size: ${Math.round(content.length / 1024)}KB`);

        try {
            // Create parent directories
            await fs.mkdir(path.dirname(filePath), { recursive: true });
            await fs.writeFile(filePath, content);
            console.log(`[LocalAdapter] ✅ File written successfully: ${filePath}`);
        } catch (error) {
            console.error(`[LocalAdapter] ❌ Failed to write file: ${objectKey}`);
            console.error(`[LocalAdapter] Error:`, error.message);
            throw new Error(`Failed to write file: ${objectKey} - ${error.message}`);
        }
    }

    async fileExists(objectKey) {
        const filePath = this._resolvePath(objectKey);
        try {
            const stats = await fs.stat(filePath);
            return stats.isFile();
        } catch (error) {
            return false;
        }
    }

    async getFileSize(objectKey) {
        const filePath = this._resolvePath(objectKey);
        try {
            const stats = await fs.stat(filePath);
            return stats.size;
        } catch (error) {
            if (error.code === 'ENOENT') {
                throw new Error(`File not found: ${objectKey}`);
            }
            throw new Error(`Failed to get file size: ${objectKey} - ${error.message}`);
        }
    }

    async listFiles(prefix) {
        const prefixPath = this._resolvePath(prefix);
        const files = [];

        try {
            const stats = await fs.stat(prefixPath);

            if (stats.isFile()) {
                files.push(prefix);
            } else if (stats.isDirectory()) {
                // Recursively list all files
                const walk = async (dir, baseDir) => {
                    const entries = await fs.readdir(dir, { withFileTypes: true });
                    for (const entry of entries) {
                        const fullPath = path.join(dir, entry.name);
                        if (entry.isFile()) {
                            const relativePath = path.relative(baseDir, fullPath);
                            files.push(relativePath.replace(/\\/g, '/'));
                        } else if (entry.isDirectory()) {
                            await walk(fullPath, baseDir);
                        }
                    }
                };
                await walk(prefixPath, this.basePath);
            }
        } catch (error) {
            // Return empty array if path doesn't exist
            return [];
        }

        return files;
    }

    async deleteFile(objectKey) {
        const filePath = this._resolvePath(objectKey);
        try {
            await fs.unlink(filePath);
        } catch (error) {
            if (error.code !== 'ENOENT') {
                throw new Error(`Failed to delete file: ${objectKey} - ${error.message}`);
            }
        }
    }
}
